nucliadb_vectors/src/data_point/ops_hnsw.rs

```diff
@@ -121,6 +122,38 @@ impl<'a, DR: DataRetriever> HnswOps<'a, DR> {
         candidates.truncate(k_neighbours);
         candidates
     }
+    fn select_neighbours_heuristic(
+        &self,
+        k_neighbours: usize,
+        candidates: Vec<(Address, Edge)>,
+    ) -> Vec<(Address, Edge)> {
+        let mut results = Vec::new();
+        let mut discarded = Vec::new();
+
+        // First, select the best candidates to link, trying to connect from all directions
+        // i.e: avoid linking to all nodes in a single cluster
+        for (x, sim) in candidates.iter() {
+            if results.len() == k_neighbours {
+                break;
+            }
+            // Keep if x is more similar to the new node than it is similar to other results
+            // i.e: similarity(x, new) > similarity(x, y) for all y in result
+            let check = results.iter().map(|&(y, _)| self.similarity(*x, y)).all(|inter_sim| *sim > inter_sim);
+            if check {
+                results.push((*x, *sim));
+            } else {
+                discarded.push((*x, *sim));
+            }
+        }
+
+        // keepPrunedConnections: keep some other connections to fill M
+        if results.len() < k_neighbours {
+            discarded.sort_unstable_by_key(|(n, d)| std::cmp::Reverse(Cnx(*n, *d)));
+            results.extend(discarded.into_iter().take(k_neighbours - results.len()));
```

Comment:
If `discarded` becomes a BinaryHeap of `std::cmp::Reverse(Cnx(*n, *d))` we could replace the if by:
```rust
while results.len() < k_neighbours {
     let Some(std::cmp::Reverse(Cnx(n, d))) = discarded.pop() else {
            return results;
     };
     results.push((n, d));
}
```
which should be better.

Reply:
I tested this. Actually, it should not have that `Reverse` in there, because this is a max-heap, so `pop()` returns the maximum value, which is already what we want, select the maximum similarity.

I micro-benchmarked both approaches and the vec + sorting is actually faster (by 1.5% or so). I think this makes sense, I expected them to be mostly equivalent, since `heap.pop()` is `O(log(n))` so popping a few elements is `O(m log(n))`. If I increase the efConstruction to 1000, it inverts, and BinaryHeap is faster (since m is smaller than n).

Times:
```
Bench               |  Vec + sort   |  BinaryHeap   |  Delta
Build full HNSW     | 24.608.256 µs | 24.993.731 µs | 98.46 %
micro: per select() |    298.419 ns |    303.408 ns | 98.35 %
```

Anyway, I'll change it to a binary heap since it's less sensitive to changes in parameters, but never underestimate a simple vector with a sort :grimacing:

Reply:
Oh no, for sure in terms of efficiency is more or less the same, the input is the list of neighbors which will be small.
I just thought the code would be easier to maintain this way